### PR TITLE
Add means to Duplicate connections from UI

### LINKED ISSUE #12401: Duplicate connections UI
In the UI it would be nice to duplicate the selected connection, in the menu "With Selected". 

The copy would be set with the same name plus some suffix (_copy, _1, whatever) and it is useful for when you have a connection with all equal fields except something in a particular one.

### PR DESCRIPTION
**Pull request by pateash**
closes: #12401

---
**Description**

* This change allows users to duplicate connections via UI.
* user can also create duplicates of multiple connections.
* A descriptive flash message will be shown regarding the success or error in any of the connection duplication.

### CODE REVIEW

**Comment by pateash** [comment]
image

image

**Comment by pateash** [comment]
It will show a user-friendly error in case of name conflict ( unique integrity)

image

**Comment by jhtimmins** [comment]
@ryanahamilton @ashb Bumping y'all on this since you're code owners here.

I think the code looks solid (nice job on the string update logic @pateash). My only nitpick is if `_Copy` should be uppercased. If connections typically have snakecase titles, it'd be cleaner to go lowercase (`_copy`). But if no one else minds then I'm fine with it.

**Comment by xinbinhuang** [comment]
I have the same feeling as @jhtimmins : should it be snakecase `_copy` instead?

Another thing is: can you also a test case to cover the duplicate logic?

File: tests/www/test_views.py [hunk 1]
```diff
@@ -270,6 +270,73 @@ def test_prefill_form_null_extra(self):
         cmv.prefill_form(form=mock_form, pk=1)
 
 
+class TestMultiDuplicateConnectionModelView(TestBase):
```
Comment (ashb) [inline tests/www/test_views.py]: Please don't create a new test class -- instead add methods to the existing TestConnectionModelView

File: tests/www/test_views.py [hunk 2]
```diff
@@ -270,6 +270,73 @@ def test_prefill_form_null_extra(self):
...
+        expected_result = {
+            'google_cloud_default',
+            'google_cloud_default_Copy(1)',
+            'mongodb',
+            'mongodb_Copy(1)',
+            'mysql_connection',
+            'postgres_connection_Copy(1)',
+            'postgres_connection_Copy(2)',
+        }
+        response = {conn[0].strip() for conn in self.session.query(Connection.conn_id).all()}
```
Comment (ashb) [inline tests/www/test_views.py]:
```suggestion
        response = {conn[0] for conn in self.session.query(Connection.conn_id).all()}
```

strip shouldn't be needed.

File: tests/www/test_views.py [hunk 3]
```diff
@@ -561,15 +626,15 @@ def test_health(self):
             BaseJob.job_type == 'SchedulerJob',
             BaseJob.state == 'running',
             BaseJob.latest_heartbeat == last_scheduler_heartbeat_for_testing_1,
-        ).delete()
+            ).delete()
```
Comment (ashb) [inline tests/www/test_views.py]: Please avoid making formatting only changes to lines you don't otherwise touch, as it makes the git history messier.

**Review by ashb — CHANGES_REQUESTED** [review]
As others have said, we should have it as `_copy` not `_Copy`.

Additionally, it should be `copy1` not `copy(1)` so it can be set via env vars if _has_ to. (You can't put `()` in env var names, so `AIRFLOW_CONN_FOO_COPY(1)` would not be possible

File: airflow/www/views.py [hunk 4]
```diff
@@ -3012,6 +3015,57 @@ def action_muldelete(self, items):
...
+    @auth.has_access(
+        [
+            (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
+        ]
+    )
+    def action_mulduplicate(self, connections, session=None):
+        """Duplicate Multiple connections"""
+        for selected_conn in connections:
+            new_conn_id = selected_conn.conn_id
+            matches = re.findall(r"_Copy\((\d+)\)", selected_conn.conn_id)
```
Comment (uranusjr) [inline airflow/www/views.py]: Is it intended to find `_CopyX` *anywhere*? Since the logic below only always adds a suffix, I would expect this to only check for a suffix as well.

File: tests/www/test_views.py (hunk 3 above)
Comment (pateash) [reply tests/www/test_views.py]: sorry @ashb, Looks like my IDE does that automatically.
I will fix these.

Comment (ashb) [reply tests/www/test_views.py]: @pateash No problem, I know how it happens, it just makes my life as a release manager much harder :)

File: airflow/www/views.py (hunk 4 above)
Comment (pateash) [reply airflow/www/views.py]: This is a good point,
In case of something like **abc_copy1_xyz**,  we will currently end up doing **abc_copy2_xyz**, but **abc_copy1_xyz_copy1** should be the accepted behaviour.

File: airflow/www/views.py [hunk 5]
```diff
@@ -3012,6 +3015,57 @@ def action_muldelete(self, items):
...
+    @auth.has_access(
+        [
+            (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
+        ]
+    )
+    def action_mulduplicate(self, connections, session=None):
+        """Duplicate Multiple connections"""
+        for selected_conn in connections:
+            new_conn_id = selected_conn.conn_id
+            matches = re.findall(r"_copy(\d+)$", selected_conn.conn_id)
```
Comment (uranusjr) [inline airflow/www/views.py]: Since there can only be one match here, it’s probably better to do

```python
match = re.search(...)
if match is not None:
    ...
```

instead.

File: airflow/www/views.py [hunk 6]
```diff
@@ -3012,6 +3015,57 @@ def action_muldelete(self, items):
...
+            if matches:
+                # replacing the last _Copy(Num) with _Copy(Num+1)
```
Comment (uranusjr) [inline airflow/www/views.py]: This comment is out of date.

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Comment by pateash** [comment]
> As others have said, we should have it as `_copy` not `_Copy`.
> 
> Additionally, it should be `copy1` not `copy(1)` so it can be set via env vars if _has_ to. (You can't put `()` in env var names, so `AIRFLOW_CONN_FOO_COPY(1)` would not be possible

Thanks All for valuable comments,
As per suggestions, I have used **_copyN**  Syntax for duplication,
also test has been added.

File: airflow/www/views.py (hunk 5 above)
Comment (pateash) [reply airflow/www/views.py]: yeah, agree as we are only matching the last occurence.
**re.search()** will be better.

File: airflow/www/views.py (hunk 6 above)
Comment (pateash) [reply airflow/www/views.py]: removed

File: airflow/www/views.py [hunk 7]
```diff
@@ -3012,6 +3015,56 @@ def action_muldelete(self, items):
...
+    )
+    def action_mulduplicate(self, connections, session=None):
+        """Duplicate Multiple connections"""
+        for selected_conn in connections:
+            new_conn_id = selected_conn.conn_id
+            matches = re.search(r"_copy(\d+)$", selected_conn.conn_id)
+            if matches:
+                new_conn_id = selected_conn.conn_id.replace(
+                    f'_copy{matches.group(1)}', f'_copy{int(matches.group(1)) + 1}'
+                )
```
Comment (uranusjr) [inline airflow/www/views.py]:
```suggestion
                conn_id_prefix = selected_conn.conn_id[:matches.start()]
                new_conn_id = f"{conn_id_prefix}_copy{int(matches.group(1)) + 1}"
```

The current implementation would fail in edge cases like `conn_id_copy1_copy1` (incorrectly replacing the first `copy1`).

Also the variable should probably be named `match` (singular) instead.

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/www/views.py (hunk 7 above)
Comment (pateash) [reply airflow/www/views.py:3038]: @uranusjr, 
it won't fail in any cases as we are only matching the last occurence in the regex.
**r"_copy(\d+)$"**

Comment (pateash) [reply airflow/www/views.py]: I agree,
that the name of variable should be singular.

File: airflow/www/views.py [hunk 8]
```diff
@@ -3012,6 +3015,55 @@ def action_muldelete(self, items):
...
+            try:
+                session.add(dup_conn)
+                session.commit()
+                flash(f"Connection {new_conn_id} added successfully.", "success")
+            except IntegrityError:
+                flash(
+                    f"Connection {new_conn_id} can't be added. Integrity error, probably unique constraint.",
+                    "warning",
+                )
+                session.rollback()
```
Comment (uranusjr) [inline airflow/www/views.py:3228]: Some thoughts on the current approach inserting the duplicated connections:

* This commits each duplicated connection separately, so any failures would not affect other duplications. I’m not sure I’d want this as a user (probably subjective).
* The error message is not very helpful either. We could make a bit more effort by querying whether there’s an existing connection to the ID, but that could be subject to race conditions as well? (I think some database-level configuration can avoid this?)
* Combining the previous two points, it may be possible to first connect all the new IDs, query whether any of them exist, and fail all of them cause conflicts or fail to insert.
* If there’s more than one failures, the user may get a bunch of flashes which may be difficult to read? But maybe that’s not really an issue in practice.

Comment (pateash) [reply airflow/www/views.py:3228]: @uranusjr,
I think the current approach is better, as all the connections are independent to each other so they will get independent status. 
Also, from an error message point of view, We are only catching integrity error, so we are providing the same error.
A similar approach is already in place in add() and edit() functions for the Connection Model.

File: tests/www/views/test_views_connection.py [hunk 9]
```diff
@@ -54,3 +54,46 @@ def test_prefill_form_null_extra():
...
+        conn_type='FTP',
+        description='Postgres',
+        host='localhost',
+        schema='airflow',
+        port=3306,
+    )
+    with create_session() as session:
+        session.query(Connection).delete()
+        session.add_all([conn1, conn2, conn3])
+        session.commit()
```
Comment (uranusjr) [inline tests/www/views/test_views_connection.py:85]: This part above should be a fixture since it’s test setup, not the actual test.

File: tests/www/views/test_views_connection.py [hunk 10]
```diff
@@ -54,3 +54,46 @@ def test_prefill_form_null_extra():
...
+
+    mock_form = mock.Mock()
+    mock_form.data = {"action": "mulduplicate", "rowid": [conn1.id, conn3.id]}
+    resp = admin_client.post('/connection/action_post', data=mock_form.data, follow_redirects=True)
```
Comment (uranusjr) [inline tests/www/views/test_views_connection.py]: Is `mock_form` needed? It doesn’t seem to be used anywhere.

It seems only this is needed:

```python
data = {"action": "mulduplicate", "rowid": [conn1.id, conn3.id]}
resp = admin_client.post('/connection/action_post', data=data, follow_redirects=True)
```

Comment (pateash) [reply tests/www/views/test_views_connection.py]: agree

File: tests/www/views/test_views_connection.py (hunk 9 above)
Comment (pateash) [reply tests/www/views/test_views_connection.py:85]: Yeah, But this is very specific to this individual test case.
if we add a fixture, we should use this in at least 2 places.

Comment (uranusjr) [reply tests/www/views/test_views_connection.py:85]: I don’t agree with the distinction. IMO test setup should be test setup, and the test function should only contain the test itself. This makes tests easier to debug against the test framework; a test setup error shows up as `E` instead of `F` in pytet, for example. It also categories run durations differently (time spent on test setup is marked with “setup” in the summary) and makes it easier to detect slow tests.

But I guess you can argue that’s subjective.

**Comment by pateash** [comment]
> As others have said, we should have it as `_copy` not `_Copy`.
> 
> Additionally, it should be `copy1` not `copy(1)` so it can be set via env vars if _has_ to. (You can't put `()` in env var names, so `AIRFLOW_CONN_FOO_COPY(1)` would not be possible

done

**Review by uranusjr — APPROVED** [review]
Nice!

**Comment by potiuk** [comment]
Can you please rebase and solve conflicts @pateash ?

**Comment by pateash** [comment]
> Can you please rebase and solve conflicts @pateash ?

Thanks @potiuk , done.

**Review by potiuk — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by jhtimmins** [comment]
@ashb can you check whether your requested changes have been addressed?

File: airflow/www/views.py [hunk 11]
```diff
@@ -3113,6 +3116,55 @@ def action_muldelete(self, items):
...
+    @action(
+        'mulduplicate',
+        'Duplicate',
+        'Are you sure you want to duplicate the selected connections?',
+        single=False,
+    )
+    @provide_session
+    @auth.has_access(
+        [
+            (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
```
Comment (ashb) [inline airflow/www/views.py:3192]:
```suggestion
            (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
            (permissions.ACTION_CAN_READ, permissions.RESOURCE_CONNECTION),            
```

Although it's unlikely you'd have create and not read, this should also have read permissions.

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/www/views.py (hunk 11 above)
Comment (pateash) [reply airflow/www/views.py:3192]: sure make sense.

**Review by uranusjr — APPROVED** [review]

**Review by ashb — APPROVED** [review]

**Comment by jyotsa09** [comment]
There is a scenario where I feel this is not working as expected.
- When I have created a connection with name **conn_test_copy1**  and then I have tried to duplicate this connection. It actually created a connection with name as **conn_test_copy2** instead of **conn_test_copy1_copy1**.

**Comment by pateash** [comment]
> There is a scenario where I feel this is not working as expected.
this is expected behaviour, you should not have any connection with _copy<X> in the end,
it's just a way to create duplicate, after duplication you are supposed to rename the connection.

**Comment by uranusjr** [comment]
We should detect this and show a better error message. @jyotsa09 Could you file an issue for this?
